tokio/src/runtime/time/entry.rs

```diff
@@ -162,26 +157,31 @@ impl StateCell {
         }
     }
 
-    /// Marks this timer as being moved to the pending list, if its scheduled
-    /// time is not after `not_after`.
+    /// Marks this timer firing, if its scheduled time is not after `not_after`.
     ///
     /// If the timer is scheduled for a time after `not_after`, returns an Err
     /// containing the current scheduled time.
     ///
     /// SAFETY: Must hold the driver lock.
-    unsafe fn mark_pending(&self, not_after: u64) -> Result<(), u64> {
+    unsafe fn mark_firing(&self, not_after: u64) -> Result<(), u64> {
         // Quick initial debug check to see if the timer is already fired. Since
         // firing the timer can only happen with the driver lock held, we know
         // we shouldn't be able to "miss" a transition to a fired state, even
         // with relaxed ordering.
         let mut cur_state = self.state.load(Ordering::Relaxed);
-
         loop {
+            // This entry is in the `guarded_list`, so it can not be removed
+            // from the entry list with the same `level` and `slot`.
+            // Because its state is STATE_DEREGISTERED, it has been fired.
+            if cur_state == STATE_DEREGISTERED {
+                break Err(cur_state);
+            }
```

Comment:
I don't understand this comment. How do you know that it's in the guarded list? And why is that a problem?

Reply:
Only the Entry which is `pop_back` ed from guarded_list can be `mark_firing`. So we can ensure this.

Reply:
If we fail to mark a Entry as `STATE_FIRING`, then we get a problem.

This entry has been fired, so we fail to fire it again.

Reply:
The code seems reasonable, I just found the comment confusing. Why is it important that it can't be removed from the entry list?

Reply:
This comment is actually my explanation of why the state value is STATE_FIRING here.

Because it can not be removed, but can be fired, so we meet this Entry here.

Please see the `remove` method of the Wheel.

Reply:
Or we can make this more clear by adding some comments.